kvdb-rocksdb/src/lib.rs

```diff
@@ -152,47 +152,53 @@ impl CompactionProfile {
 pub struct DatabaseConfig {
 	/// Max number of open files.
 	pub max_open_files: i32,
-	/// Memory budget (in MiB) used for setting block cache size, write buffer size.
-	pub memory_budget: Option<usize>,
+	/// Memory budget (in MiB) used for setting block cache size, 
+	/// write buffer size for each column.
+	pub memory_budget: Vec<Option<usize>>,
 	/// Compaction profile
 	pub compaction: CompactionProfile,
-	/// Set number of columns
-	pub columns: Option<u32>,
 }
 
 impl DatabaseConfig {
 	/// Create new `DatabaseConfig` with default parameters and specified set of columns.
 	/// Note that cache sizes must be explicitly set.
 	pub fn with_columns(columns: Option<u32>) -> Self {
 		let mut config = Self::default();
-		config.columns = columns;
+		config.memory_budget.resize(columns.unwrap_or(0) as usize, None);
 		config
 	}
 
 	pub fn memory_budget(&self) -> usize {
-		self.memory_budget.unwrap_or(DB_DEFAULT_MEMORY_BUDGET_MB) * MB
+		if self.memory_budget.is_empty() {
+			return DB_DEFAULT_MEMORY_BUDGET_MB * MB;
```

Comment:
If both fields are not set (memory_budget and memory_budget_per_col), all columns will have DB_DEFAULT_MEMORY_BUDGET_MB size and overall size will be the same DB_DEFAULT_MEMORY_BUDGET_MB. Is it ok?

Reply:
See the PR description:
> One of the things I'm not happy about in this PR, is that it doesn't allow to configure memory budget in case of 0 non-default columns (i.e. Database::open_default()), but it seems like we always use default settings for that case in parity-ethereum.

Reply:
I was thinking of making memory_budget an enum like
```rust
enum MemoryBudget {
    DefaultOnly(Option<usize>),
    PerColumn(Vec<Option<usize>>),
```
but that would complicate the API and we don't seem to use it anyway.

Reply:
Didn't understand properly your note in the description :-( 

Why not to use the current logic (memory_budget_for_col = total \ amount_of_cols), if exact values for columns are not set? It's a good heuristic. It's a just a total miss for Parity Ethereum. So in this case Parity Ethereum should use proper defaults for columns.

Reply:
> Why not to use the current logic (memory_budget_for_col = total \ amount_of_cols), if exact values for columns are not set?

I am also curious about this, can you elaborate @ordian ?

Reply:
(this was discussed privately, so I didn't reply)
The whole point of this change is to move the logic of dividing the the whole budget to the user. So e.g. parity-ethereum client will divide the budget and provide a config with `vec<90%, 10%, ... >` memory budget to kvdb-rocksdb. Does that make sense?

Reply:
If some parts are not set, it's not clear what the whole budget is how are we supposed to distribute it, so the fixed default amount is used.

Reply:
> So e.g. parity-ethereum client will divide the budget and provide a config with `vec<90%, 10%, ... >` memory budget to kvdb-rocksdb. Does that make sense?

I actually though the user were to provide absolute sizes; proportions make more sense. Docs definitely needed.

Reply:
No, I meant absolute sizes, sorry for confusion. Since the whole budget is not passed, it's the only info we have.